server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/csv/IRVHeadersParser.java

```diff
@@ -0,0 +1,128 @@
+import au.org.democracydevelopers.corla.model.vote.IRVPreference;
+import org.apache.commons.csv.CSVRecord;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import us.freeandfair.corla.model.Choice;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Utils for parsing IRV headers as part of DominionCVRExportParser.
+ * Main methods:
+ * - validateIRVPreferenceHeaders validates a whole collection as part of a CSV header, to check
+ *   that they are consistent with the expected pattern.
+ * - generateAllIRVPreferences produces the expected pattern of preferences, given the list of
+ *   choices and the number of ranks allowed.
+ */
+public class IRVHeadersParser {
+  /**
+   * Class-wide logger
+   */
+  private static final Logger LOGGER = LogManager.getLogger(IRVHeadersParser.class);
+
+  /**
+   * Checks whether a CSVRecord contains a sequence of choices as expected for IRV - it should be
+   * a sequence of name(rank) choices, starting with preference 1 and going up to maxRank,
+   * and for each rank, it should contain all the same choices in the same order. For example,
+   * Alice(1), Bob(1), Chuan(1), Alice(2), Bob(2), Chuan(2), Alice(3), Bob(3), Chuan(3) is valid,
+   * but any rearrangement of those items, or anything that omits some rank or candidate included
+   * elsewhere, is invalid.
+   * @param theLine the CSV line (basically a list of Strings)
+   * @param startIndex the first index for this contest's choices.
+   * @param numChoices the number of choices (candidates).
+   * @param maxRank The number of ranks allowed (which matches the number expected in the csv).
+   * @throws IRVParsingException if either an individual choice can't be parsed as name(rank), or
+   *                             the overall collection of choices doesn't fit the pattern above.
+   * This works by stepping through the 1st-rank choices, checking that the name has not been
+   * repeated, and then checking that the later-ranked versions of the same choice (located
+   * (rank_being_checked-1)*numChoices further along the line) have the same name and the correct
+   * rank.
+   */
+  public static void validateIRVPreferenceHeaders(CSVRecord theLine, int startIndex, int numChoices,
+                                                  int maxRank) throws IRVParsingException {
+    final String prefix = "[validateIRVChoiceHeaders] ";
+    final String errorMsg = "Invalid IRV choices header: ";
+
+    final Set<String> names = new HashSet<>();
+
+    try {
+      // Iterate over the first-preference choices, expected to have distinct names.
+      for (int i = startIndex; i < numChoices + startIndex; i++) {
+        IRVPreference irv1 = new IRVPreference(theLine.get(i));
+
+        // A repeated name, or a preference other than 1, is an error.
+        if (!names.add(irv1.candidateName) || irv1.rank != 1) {
```

Comment:
Log throwing of exception here. Can more detail be added to the error message eg. the details of the IRVpreference that caused the error (I know the line itself will be added later in the catch)?

Reply:
Have added the actual preferences that caused the error into the error message.
The exception is caught at the bottom of that function and logged there, which sort of saved an extra 2 lines, but if it's confusing then I can easily put the logging explicitly into each case.